tmpl/blog/functoria.md

```diff
@@ -0,0 +1,153 @@
+```
+
+Here, we defined both a long option `--lang` and a short one `-l` (the format is similar to the one used by [Cmdliner][cmdliner]).
+In the unikernel, the value is retrieved with `Bootvar_gen.language ()`.
+
+The option is also documented in the `--help` option for both `mirage configure` (at configure time) and `./my_unikernel` (at startup time).
+
+```
+       -l VAL, --lang=VAL (absent=en)
+           The default language for the unikernel.
+```
+
+[cmdliner]: http://erratique.ch/software/cmdliner
+
+A simple example of a unikernel with a key is available in [mirage-skeleton][] in the `hello` directory.
+
+### Switching implementation
+
+We can do much more with keys: we can use them to switch devices at configure time.
+To illustrate, let us take the example of a dynamic storage: we want to choose between a block device and a crunch device with a command line option.
+In order to do that, we must first define a boolean key:
+
+```
+let fat_key =
+  let doc = Key.Doc.create
+      ~doc:"Use a fat device if true, crunch otherwise." [ "fat" ]
+  in
+  Key.create ~doc ~stage:`Configure ~default:false "fat" Key.Desc.bool
+```
+
+We can use the `if_impl` combinator to choose between two devices depending on the value of the key.
+
+```
+let dynamic_storage =
+  if_impl (Key.value fat_key)
+    (kv_ro_of_fs my_fat_device)
+    (my_crunch_device)
+```
+
+We can now use this device as a normal storage device of type `kv_ro impl`! The key is also documented in `mirage configure --help`:
+
+```
+       --fat=VAL (absent=false)
+           Use a fat device if true, crunch otherwise.
+```
+
+It is also possible to compute on keys before giving them to `if_impl`, combining multiple keys in order to compute a value, and so on. The documentation for the API is located in the `Mirage.Key` module and various examples are available in [mirage][] and [mirage-skeleton][].
+
+Switching keys opens various possibilities, for example a `generic_stack` combinator is now implemented in mirage that will switch between socket stack, direct stack with DHCP and direct stack with static IP, depending on command line arguments.
+
+## Drawing unikernels
+
+All these keys and dynamic implementations make for complicated unikernels. In order to still be able to understand what is going on and how to configure our unikernels, we have a new command: `describe`.
+
+Let us consider the `console` example in [mirage-skeleton][]:
+
+```
+open Mirage
+
+let main = foreign "Unikernel.Main" (console @-> job)
+let () = register "console" [main $ default_console]
+```
+
+This is fairly straightforward: we define a `Unikernel.Main` functor using a console and we
+instantiate it with the default console. If we execute `mirage describe --dot` in this directory, we will get the following output.
+
+[![A console unikernel](../graphics/dot/console.svg "My little unikernel")](../graphics/dot/console.svg)
+
+As you can see, there is already quite a few things going on!
```

Comment:
Either "there are already quite a few" or "there is already quite a lot" would be better, I think.